File: 2D Dynamic Programming/1289. Minimum Falling Path Sum II/1289.py
class Solution:
    def minFallingPathSum(self, grid: List[List[int]]) -> int:
        n,m=len(grid),len(grid[0])
        dp=grid[0].copy()

        for i in range(1,n):
            temp=dp.copy()
            for j in range(m):
                best=float('inf')
                for k in range(0,j):
                    best=min(best,dp[k])
                for k in range(j+1,m):
                    best=min(best,dp[k])
                temp[j]=best+grid[i][j]
            dp=temp
        return min(dp)

File: 2D Dynamic Programming/1289. Minimum Falling Path Sum II/1289.cpp
#pragma GCC optimize('0fast','unroll-loops')

class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& grid) {
        ios_base::sync_with_stdio(0);
        cin.tie(0);
        cout.tie(0);

        int n=grid.size();
        int m=grid[0].size();
        vector<int>dp(grid[0].begin(),grid[0].end());

        for(int i=1; i<n; ++i){
            vector<int>temp(dp.begin(),dp.end());
            for(int j=0; j<m; ++j){
                int best=INT_MAX;
                for(int k=0; k<m; ++k){
                    if(k!=j) best=min(best,dp[k]);
                }
            temp[j]=best+grid[i][j];
            }
            dp=temp;
        }
        return *min_element(dp.begin(),dp.end());
    }
};